src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -379,6 +408,29 @@ def _finalize_pnad_side_predicates(pnad: PartialNSRTAndDatastore) -> None:
                 side_predicates.add(atom.predicate)
         pnad.op = pnad.op.copy_with(side_predicates=side_predicates)
 
+    @staticmethod
+    def _get_pnads_with_keep_effects(
+            pnad: PartialNSRTAndDatastore) -> Set[PartialNSRTAndDatastore]:
+        """Return a new set of PNADs that include keep effects into the given
+        PNAD."""
+        # The keep effects that we want are the subset of possible keep
+        # effects which are not already in the PNAD's add effects, and
+        # whose predicates were determined to be side predicates.
+        keep_effects = {
+            eff
+            for eff in pnad.poss_keep_effects if eff not in pnad.op.add_effects
+            and eff.predicate in pnad.op.side_predicates
+        }
+        if not keep_effects:
+            return set()
+        preconditions = pnad.op.preconditions | keep_effects
+        add_effects = pnad.op.add_effects | keep_effects
+        new_pnad_op = pnad.op.copy_with(name=f"{pnad.op.name}-KEEP",
+                                        preconditions=preconditions,
+                                        add_effects=add_effects)
+        new_pnad = PartialNSRTAndDatastore(new_pnad_op, [], pnad.option_spec)
+        return {new_pnad}
```

Comment:
right now we only return one new pnad with ALL keep effects included. soon we will change this to generate combinatorially many (@wmcclinton is working on a test case for this)